sdks/go/pkg/beam/core/runtime/graphx/xlang_test.go

```diff
@@ -203,3 +203,214 @@ func TestExpandedComponents(t *testing.T) {
+				s := g.Root()
+				edge := g.NewEdge(s)
+				edge.Op = graph.External
+				edge.External = &testExternal
+
+				return []*graph.MultiEdge{edge}
+			},
+			wantExpandeds: []graph.ExpandedTransform{testExpanded},
+		},
+		{
+			name: "MultiTransforms",
+			makeEdges: func(g *graph.Graph) []*graph.MultiEdge {
+				s := g.Root()
+				edge := g.NewEdge(s)
+				edge.Op = graph.External
+				edge.External = &testExternal
+				edge2 := g.NewEdge(s)
+				edge2.Op = graph.External
+				edge2.External = &testExternal2
+
+				return []*graph.MultiEdge{edge, edge2}
+			},
+			wantExpandeds: []graph.ExpandedTransform{testExpanded, testExpanded2},
+		},
+		{
+			name: "NonExternalTransforms",
+			makeEdges: func(g *graph.Graph) []*graph.MultiEdge {
+				s := g.Root()
+				edge := g.NewEdge(s)
+				edge.Op = graph.External
+				edge.External = &testExternal
+				edge2 := g.NewEdge(s)
+				edge2.Op = graph.External
+				edge2.External = &testExternal2
+				impulse := graph.NewImpulse(g, s, []byte{1})
+				impulse2 := graph.NewImpulse(g, s, []byte{2})
+
+				return []*graph.MultiEdge{edge, edge2, impulse, impulse2}
+			},
+			wantExpandeds: []graph.ExpandedTransform{testExpanded, testExpanded2},
+		},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			g := graph.New()
+			edges := test.makeEdges(g)
+			var p pipepb.Pipeline
+			p.Components = &pipepb.Components{
+				Transforms:          make(map[string]*pipepb.PTransform),
+				Pcollections:        make(map[string]*pipepb.PCollection),
+				WindowingStrategies: make(map[string]*pipepb.WindowingStrategy),
+				Coders:              make(map[string]*pipepb.Coder),
+				Environments:        make(map[string]*pipepb.Environment),
+			}
+			mergeExpandedWithPipeline(edges, &p)
+
+			// Check that all wanted expanded components have been added to
+			// pipeline components.
+			for _, exp := range test.wantExpandeds {
+				wantComps := exp.Components.(*pipepb.Components)
+				gotComps := p.GetComponents()
+				validateComponents(t, wantComps, gotComps)
+
+				// Check that expanded transform is present. Need to search.
+				wantTransform := exp.Transform.(*pipepb.PTransform)
+				var found bool
+				for _, gotTransform := range gotComps.GetTransforms() {
+					if d := cmp.Diff(wantTransform, gotTransform, protocmp.Transform()); d == "" {
+						found = true
+						break
+					}
+				}
+				if !found {
+					t.Errorf("Pipeline components missing expected expanded transform: %v", wantTransform)
+				}
+
+				// Check that requirements are present.
+				for _, wantReq := range exp.Requirements {
+					var found bool
+					for _, gotReq := range p.GetRequirements() {
+						if wantReq == gotReq {
+							found = true
+							break
+						}
+					}
+					if !found {
+						t.Errorf("Pipeline missing expected requirement: %v", wantReq)
+					}
+				}
+			}
+		})
+	}
+}
+
+func validateComponents(t *testing.T, wantComps, gotComps *pipepb.Components) {
```

Comment:
Actually, the main reason was because the pipeline object can contain the components from multiple external transforms, so just diffing the components wouldn't work. What I needed to validate is that the components of the external transform are a subset of the pipeline components, not that they're exactly equal.

I'll add a comment explaining that, because I agree it's not immediately obvious.